designs/node-overlays.md

```diff
@@ -0,0 +1,186 @@
+# Node Overlays
+
+Node launch decisions are the output a scheduling simulation algorithm that involves satisfying pod scheduling constraints with available instance type offerings. Cloud provider implementations are responsible for surfacing these offerings to the Karpenter scheduler through the [cloudprovider.GetInstanceTypes()](https://github.com/kubernetes-sigs/karpenter/blob/37db06f4742eada19934f76e616f2b1860aca57b/pkg/cloudprovider/types.go#L59) API. This separation of responsibilities enables cloud providers to make simplifying assumptions, but limits extensibility for more advanced use cases. NodeOverlays enable customers to inject alternative assumptions into the scheduling simulation for more accurate scheduling simulations.
+
+## Use Cases
+
+### Price
+
+Each instance offering comes with a price, which is used as a minimization function in Karpenter's scheduling algorithm. However, Karpenter cannot perfectly understand all factors that influence price, such as an external vendor's fee, a pricing discount, or an external motiviation such as carbon-offset.
+
+* https://github.com/aws/karpenter-provider-aws/pull/4686
+* https://github.com/aws/karpenter-provider-aws/issues/3860
+* https://github.com/aws/karpenter-provider-aws/pull/4697
+
+### Extended Resources
+
+InstanceTypes have a set of well known resource capacities, such as CPU, memory, or GPUs. However, Kubernetes supports arbitrary [extended resources](https://kubernetes.io/docs/tasks/administer-cluster/extended-resource-node/), which cannot be known prior to Node launch. Karpenter lacks a mechanism to teach the simulation about extended resource capacities.
+
+* https://github.com/kubernetes-sigs/karpenter/issues/751
+* https://github.com/kubernetes-sigs/karpenter/issues/729
+
+### Resource Overhead
+
+* https://github.com/aws/karpenter-provider-aws/issues/5161
+
+System software such as containerd, kubelet, and the host operating system come with resource overhead like CPU and memory, which is substracted from a Node's capacity. Karpenter attempts to calculate this overhead where it is known, but cannot predict the overhead required by custom operating systems. To complicate this, the overhead often changes for each version of an operating system.
+
+## Proposal
+
+NodeOverlays are a new Karpenter API that enable customers to fine-tune the scheduling simulation for advanced use cases. NodeOverlays can be configured with a `metav1.Selector` which flexibly constrain when they are applied to a simulation. Selectors can match well-known labels like `node.kubernetes.io/instance-type` or `karpenter.sh/node-pool` and custom labels defined in a NodePool's `.spec.labels` or `.spec.requirements`. When multiple NodeOverlays match an offering for a given simulation, they are merged together using `spec.weight` to resolve conflicts.
```

Comment:
As they're NodeOverlays, does it make sense to say it just references the same as a pod NodeSelector? 

Additionally, if I'm a user, and I'm wondering if my NodeOverlays are right, will these match the node labels or the nodeclaim labels? Realistically it shouldn't matter

Reply:
I'm not grokking this